presto-main/src/main/java/com/facebook/presto/operator/scalar/ScalarFunctionImplementation.java

```diff
@@ -37,11 +41,48 @@ public ScalarFunctionImplementation(boolean nullable, List<Boolean> nullableArgu
 
     public ScalarFunctionImplementation(boolean nullable, List<Boolean> nullableArguments, MethodHandle methodHandle, Optional<MethodHandle> instanceFactory, boolean deterministic)
     {
+        this(nullable, nullableArguments, methodHandle, instanceFactory, deterministic, false, Optional.empty());
+    }
+
+    public ScalarFunctionImplementation(
+            boolean nullable,
+            List<Boolean> nullableArguments,
+            MethodHandle methodHandle,
+            Optional<MethodHandle> instanceFactory,
+            boolean deterministic,
+            boolean returnValueAsParameter,
+            Optional<Integer> returnValueSliceLength)
+    {
         this.nullable = nullable;
         this.nullableArguments = ImmutableList.copyOf(requireNonNull(nullableArguments, "nullableArguments is null"));
         this.methodHandle = requireNonNull(methodHandle, "methodHandle is null");
         this.instanceFactory = requireNonNull(instanceFactory, "instanceFactory is null");
         this.deterministic = deterministic;
+        this.returnValueAsParameter = returnValueAsParameter;
+        this.returnValueSliceLength = returnValueSliceLength;
+
+        int expectedParametersCount = methodHandle.type().parameterCount();
+        if (returnValueAsParameter) {
+            expectedParametersCount--;
+        }
+        if (instanceFactory.isPresent()) {
+            expectedParametersCount--;
+        }
+        if (methodHandle.type().parameterCount() >= 1 && methodHandle.type().parameterType(0) == ConnectorSession.class) {
+            expectedParametersCount--;
+        }
+        checkArgument(nullableArguments.size() == expectedParametersCount, "method handle parameters count mismatch");
+
+        if (returnValueAsParameter) {
+            checkArgument(!nullable, "nullable not supported for functions returning value as parameter");
+            checkArgument(getMethodHandle().type().returnType() == void.class, "methodHandle returns non-void for function returning value as parameter");
+            int returnParameterPos = instanceFactory.isPresent() ? 1 : 0;
```

Comment:
I like the fact that we force position of special parameters but I'm not sure about forcing the order of them.
Even if we force the order - I think this is invalid when there is both instanceFactory and ConnectorSession passed? If that is correct please add test to cover that for future as well. Also if we want to force the order it would be nice to check it explicitly.